appcelerator/titanium_mobile
# [6_1_X][TIMOB-24987] Android: Setting a custom theme's parent to "Theme.Titanium" no longer causes a recursive lookup.

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-24987

**Notes:**
- Modified "Theme.Titanium" to no longer use the AndroidManifest.xml <application> tag's  theme as its parent.
  * This was the cause of the recursion if the custom theme was applied to the application.
  * Fixed by reverting PR #8592 (ie: [TIMOB-19629](https://jira.appcelerator.org/browse/TIMOB-19629)) that was committed for 6.1.0.
- Fixed ListView to correctly apply its activity's theme to its rows, headers, and footers.
  * This addresses the bug for PR #8592 that I reverted (mentioned above).

**Test Case:**

1. Create a "classic" Titanium app project.
2. In this project, add the following folder: `./platform/android/res/values`
3. Under this folder, create a new "custom_theme.xml" file using the contents down below.
4. Add the below "tiapp.xml" Android settings to your project. (Note the theme reference.)
5. Use the below "app.js" code in your project.
6. Build and run the project on an Android device.
7. Verify that all button text are not in all-caps and labels are green. _(Note: This verifies that [TIMOB-19629](https://jira.appcelerator.org/browse/TIMOB-19629) is still fixed.)_
8. Open the "custom_theme.xml" file.
9. Replace XML style `parent` attribute `"@style/Theme.AppCompat"` with `"@style/Theme.Titanium"` and save.
10. Delete the project's generated "build" directory. _(This is to force a rebuild.)_
11. Build and run the project on an Android device.
12. Look at the Android log and verify that the following error message does **NOT** appear. _(This verifies that [TIMOB-24987](https://jira.appcelerator.org/browse/TIMOB-24987) is fixed.)_
- `ResourceType: Attempt to retrieve bag 0x7f050090 which is invalid or in a cycle.`

---
**custom_theme.xml**
```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
	<style name="Theme.Custom" parent="@style/Theme.AppCompat">
		<item name="android:windowBackground">@drawable/appicon</item>
		<item name="android:windowActionBar">false</item>		
		<item name="android:windowNoTitle">true</item>
		<item name="windowActionBar">false</item>
		<item name="windowNoTitle">true</item>
		<item name="android:textAllCaps">false</item>
		<item name="android:textColorTertiary">#00FF00</item>
	</style>
</resources>
```
---
**tiapp.xml**
```xml
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<application android:theme="@style/Theme.Custom"/>
		</manifest>
	</android>
</ti:app>
```
---
**app.js**
```javascript
var window = Ti.UI.createWindow(
{
	layout: "vertical",
});
var listItemTemplate =
{
	properties:
	{
		height: Ti.UI.SIZE,
	},
	childTemplates:
	[
		{
			bindId: "rowView",
			type: "Ti.UI.View",
			properties:
			{
				layout: "horizontal",
				height: Ti.UI.SIZE,
			},
			childTemplates:
			[
				{
					type: "Ti.UI.Label",
					bindId: "title",
					properties:
					{
						height: Ti.UI.SIZE,
						left: "5%",
					}
				},
				{
					type: "Ti.UI.Button",
					bindId: "button",
					properties:
					{
						title: "Click Me",
						width: Ti.UI.SIZE,
						height: Ti.UI.SIZE,
						left: "5%",
					}
				},
			],
		},
	],
};
var listSection = Ti.UI.createListSection(
{
	headerView: Ti.UI.createButton({ title: "Header Button" }),
	footerView: Ti.UI.createButton({ title: "Footer Button" }),
});
for (var index = 1; index <= 4; index++) {
	listSection.appendItems(
	[
		{
			title: { text: "Row " + index.toString() },
		},
	]);
}
var listView = Ti.UI.createListView(
{
	templates: { 'listItemTemplate': listItemTemplate },
	defaultItemTemplate: "listItemTemplate",
});
listView.sections = [listSection];
window.add(listView);
window.open();
```
